src/oss/langchain-messages.mdx

```diff
@@ -2,14 +2,488 @@
+print(f"Response metadata: {response.response_metadata}")
+```
+
+### Response Metadata
+
+AI messages include metadata from the provider with token usage, model details, and generation information:
+
+```python
+response = model.invoke("Explain quantum computing")
+metadata = response.response_metadata
+
+print(f"Model: {metadata.get('model_name')}")
+print(f"Tokens: {metadata.get('token_usage', {}).get('total_tokens', 0)}")
+print(f"Finish reason: {metadata.get('finish_reason')}")
+```
+
+Common fields include `model_name`, `finish_reason`, `id`, and `token_usage` (format varies by provider).
+
+### Tool Calling Responses
+
+When models make tool calls, they're included in the AI message:
+
+TODO schema switcher
+
+```python
+from pydantic import BaseModel, Field
+
+class GetWeather(BaseModel):
+    """Get weather for a location"""
+    location: str = Field(..., description="City and state")
+
+llm_with_tools = llm.bind_tools([GetWeather])
+response = llm_with_tools.invoke([HumanMessage("What's the weather in Paris?")])
+
+# Access tool calls
+if response.tool_calls:
+    for tool_call in response.tool_calls:
+        print(f"Tool: {tool_call['name']}")
+        print(f"Args: {tool_call['args']}")
+        print(f"ID: {tool_call['id']}")
+```
+
+### Streaming and Chunks
+
+During streaming, you receive `AIMessageChunk` objects that accumulate metadata:
+
+```python
+# Stream response and collect metadata
+chunks = []
+for chunk in llm.stream([HumanMessage("Write a poem")]):
+    chunks.append(chunk)
+    print(chunk.content, end="", flush=True)
+
+# TODO show how they are auto-summed in new pattern
+```
+
 ## Tool Message
 
+Tool messages contain the results of tool executions and are used to pass data back to models after external processing.
+
+TODO link out to tool calling conceptual guide
+
+### Basic Tool Messages
+
+```python
+from langchain_core.messages import ToolMessage, AIMessage
+
+# After a model makes a tool call
+ai_message = AIMessage(
+    tool_calls=[{
+        "name": "get_weather",
+        "args": {"location": "San Francisco"},
+        "id": "call_123"
+    }]
+)
+
+# Execute tool and create result message
+weather_result = "Sunny, 72°F"
+tool_message = ToolMessage(
+    content=weather_result,
+    tool_call_id="call_123"  # Must match the call ID
+)
+
+# Continue conversation
+messages = [
+    HumanMessage("What's the weather in San Francisco?"),
+    ai_message,  # Model's tool call
+    tool_message,  # Tool execution result
+]
+response = model.invoke(messages)  # Model processes the result
+```
+
+### Tool Message Structure
+
+```python
+tool_message = ToolMessage(
+    content="Operation completed successfully",
+    tool_call_id="call_123",  # Links to specific tool call
+    name="calculator",  # Optional tool name
+    artifact={"raw_data": [1, 2, 3]}  # Additional data not sent to model
+    # TODO link out to artifact guide ref
+)
+```
+
+### Error Handling in Tool Messages
```

Comment:
dont think this is relevant for this? this is probably more relevant for Tool calling